### PR TITLE
[CS] Some SyntacticElement refactorings

### PR DESCRIPTION
**Pull request by hamishknight**
Split off some refactorings from the if expression work. This should be NFC.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test Linux

File: lib/Sema/BuilderTransform.cpp [hunk 1]
```diff
@@ -364,68 +364,70 @@ class BuilderClosureVisitor
     }
   }
 
-  VarDecl *visitBraceStmt(BraceStmt *braceStmt) {
-    SmallVector<Expr *, 4> expressions;
+  void visitBraceElement(ASTNode node, SmallVectorImpl<Expr *> &expressions) {
```
Comment (xedin) [inline lib/Sema/BuilderTransform.cpp:367]: Let’s not do this as BuilderClosureTransform and all relative logic is going away completely soon - https://github.com/apple/swift/pull/60419

File: lib/Sema/CSSyntacticElement.cpp [hunk 2]
```diff
@@ -358,16 +358,24 @@ class SyntacticElementConstraintGenerator
   friend StmtVisitor<SyntacticElementConstraintGenerator, void>;
 
   ConstraintSystem &cs;
-  AnyFunctionRef context;
+  DeclContext *dc;
```
Comment (xedin) [inline lib/Sema/CSSyntacticElement.cpp:361]: This is actually moving in opposite direction (it used to be DeclContext originally) but we’d like to support `TapExpr` as a context as well - https://github.com/apple/swift/pull/61746

File: lib/Sema/CSBindings.cpp [hunk 3]
```diff
@@ -1065,6 +1065,14 @@ bool BindingSet::favoredOverDisjunction(Constraint *disjunction) const {
   return !involvesTypeVariables();
 }
 
+bool BindingSet::favoredOverConjunction(Constraint *conjunction) const {
+  if (CS.shouldAttemptFixes() && isHole()) {
+    if (forClosureResult() || forGenericParameter())
```
Comment (xedin) [inline lib/Sema/CSBindings.cpp:1070]: Why is this important in diagnostic mode?

**Review by xedin — CHANGES_REQUESTED** [review]

File: lib/Sema/CSBindings.cpp (hunk 3 above)
Comment (hamishknight) [reply lib/Sema/CSBindings.cpp:1070]: No idea, I'm just preserving the original logic :)

File: lib/Sema/CSSyntacticElement.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/CSSyntacticElement.cpp:361]: Makes sense, shall I pull in SyntacticElementContext minus `TapExpr` for this PR?

Comment (xedin) [reply lib/Sema/CSSyntacticElement.cpp:361]: Sure, it should be cherry-pickable.

File: lib/Sema/CSBindings.cpp (hunk 3 above)
Comment (xedin) [reply lib/Sema/CSBindings.cpp:1070]: I should be trying to remember then :)

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/CSSyntacticElement.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/CSSyntacticElement.cpp]: How does this look?

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by xedin — APPROVED** [review]

**Pull request merged**